src/Revisionable/Document/Repository/RevisionRepository.php

```diff
@@ -0,0 +1,182 @@
+
+        $documentId = (string) $documentWrapper->getIdentifier(false, true);
+        $documentClass = $documentWrapper->getMetadata()->getName();
+
+        $qb = $this->createQueryBuilder();
+        $qb->field('revisionableId')->equals($documentId);
+        $qb->field('revisionableClass')->equals($documentClass);
+        $qb->sort('version', 'DESC');
+
+        return $qb->getQuery()->getIterator()->toArray();
+    }
+
+    /**
+     * Reverts the given document to the requested version, restoring all versioned fields to the state of that revision.
+     *
+     * Callers to this method will need to persist and flush changes to the document.
+     *
+     * @param T            $document
+     * @param positive-int $version
+     *
+     * @throws UnexpectedValueException
+     */
+    public function revert(object $document, int $version = 1): void
+    {
+        $documentWrapper = new MongoDocumentWrapper($document, $this->getDocumentManager());
+
+        $documentMetadata = $documentWrapper->getMetadata();
+        $documentId = (string) $documentWrapper->getIdentifier(false, true);
+        $documentClass = $documentMetadata->getName();
+
+        $qb = $this->createQueryBuilder();
+        $qb->field('revisionableId')->equals($documentId);
+        $qb->field('revisionableClass')->equals($documentClass);
+        $qb->field('version')->lte($version);
+        $qb->sort('version', 'ASC');
+
+        $revisions = $qb->getQuery()->getIterator()->toArray();
+
+        if ([] === $revisions) {
+            throw new UnexpectedValueException(sprintf('Could not find any revisions for version %d of document %s.', $version, $documentClass));
+        }
+
+        $data = [[]];
+
+        while ($revision = array_shift($revisions)) {
+            $data[] = $revision->getData();
+        }
+
+        $data = array_merge(...$data);
+
+        $this->fillDocument($document, $data);
+    }
+
+    /**
+     * Fills a document's versioned fields with the given data
+     *
+     * @param T                    $document
+     * @param array<string, mixed> $data
+     */
+    protected function fillDocument(object $document, array $data): void
+    {
+        $documentWrapper = new MongoDocumentWrapper($document, $this->getDocumentManager());
+
+        $documentMeta = $documentWrapper->getMetadata();
+
+        assert($documentMeta instanceof ClassMetadata);
+
+        $config = $this->getListener()->getConfiguration($this->getDocumentManager(), $documentMeta->getName());
+        $fields = $config['versioned'];
+
+        foreach ($data as $field => $value) {
+            if (!in_array($field, $fields, true)) {
+                continue;
+            }
+
+            $mapping = $documentMeta->getFieldMapping($field);
+
+            // Fill the embedded document
+            if ($documentWrapper->isEmbeddedAssociation($field)) {
+                if (!empty($value)) {
```

Comment:
What are the values allowed for `$value`? As `empty()` will return `true` for falsy values like `false` and `0`, I'd try to avoid using this function in favor of a stricter comparison.

Reply:
I'd have to inspect what's actually getting stored in different scenarios to see what's going on here, but this is largely copy/pasted from the loggable repository and I'm not a heavy MongoDB user so I'm not sure if stuff that'll cause issues with `empty()` is actually getting logged.

Reply:
```sh
1 "isEmbeddedAssociation"
2 array:2 [
  "name" => "John Doe"
  "email" => "john@doe.com"
]
```

The only time we enter this branch with the current test fixtures it has an array with the embedded document's revisioned data.  Given that we recursively get into another `fillDocument()` method call and the `$value` is passed to the typehinted `array $data` argument, we might be able to go to a `$value !== []` check.  This will need more testing though to validate the assumption.

Reply:
I added more tests here and so far it's either an array or null.  So switched to an explicit `null !== $value` check for the moment, but if this causes problems then this can be loosened up.